README.md
# better-tactics

Better Tactics is a Chess tactics trainer that uses the concept of <a href="https://en.wikipedia.org/wiki/Spaced_repetition">Spaced Repetition</a> to help you master chess tactics. The idea is to help you improve your tactical pattern recognition and gain experience calcualting tactics, by repeating puzzles you've seen previously over time. Puzzles you complete will get queued up for review daily, and puzzles you find hard will be repeated more
frequently, while puzzles you find easy will be repeated far less often to make the process more efficient.

<!-- Screenshot gallery -->
<a href="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/1.png">
  <img src="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/1_thumb.jpg">
</a>
<a href="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/2.png">
  <img src="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/2_thumb.jpg">
</a>
<a href="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/3.png">
  <img src="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/3_thumb.jpg">
</a>
<a href="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/4.png">
  <img src="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/4_thumb.jpg">
</a>
<a href="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/5.png">
  <img src="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/5_thumb.jpg">
</a>
<a href="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/6.png">
  <img src="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/6_thumb.jpg">
</a>

# Set up

There are three options to get it up and running:
* Prebuilt binaries (windows only) on <a href="https://github.com/catchouli/better_tactics/releases/latest">the releases page</a>
* From source, using cargo\*: `cargo run --release`
* Using Docker, a prebuilt image is available at: <a href="https://hub.docker.com/r/catchouli/better-tactics">catchouli/better-tactics</a>. See docker-compose.yml for an example.

Npm needs to be installed to build the frontend, but it is automatically invoked when building with cargo. If you don't have it, then install Node from: https://nodejs.org/en/download

\* Note: For a standalone/portable build use the release build, as it compiles the static assets in, but the debug build references them from the ./assets directory.

Once it says it's serving the site go to e.g. http://localhost:3030 to use the app. The lichess puzzles database will be automatically downloaded and imported in the background the first time you run the application. If you don't let it finish, it'll start over again the next time you run it. The layout also has support for mobile devices; if you change BIND_INTERFACE to 0.0.0.0 (bind on all interfaces, see CONFIG.md and .env), you will also be able to access it from a phone browser over your local network.

Configuration variables can be set using environment variables or a .env file. See the included <a href="https://github.com/catchouli/better_tactics/blob/main/.env">.env file</a> for an example, and <a href="https://github.com/catchouli/better_tactics/blob/develop/CONFIG.md">CONFIG.md</a> for information about all available config variables.

# How to use

To begin, start by practicing some new tactics puzzles on the Practice page. There, a new puzzle of an appropriate difficulty will be picked, and you'll be given the chance to study the puzzle and then identify and input the forcing line which results in checkmate, wins you material, or gains you some other advantage. After you've played some puzzles, you'll have daily reviews, which you can do on the Review page.

When you complete a puzzle, you'll be given the option to score it by how difficult you found it. The Spaced Repetition algorithm will then queue up the puzzle to be shown again to you in the future, depending on the score you picked.

If you get really stuck on a puzzle, you can click the 'Analyse' link next to the puzzle to be be taken to the Analysis Board on <a href="https://lichess.org">lichess.org</a>, where you can analyse the puzzle using an engine to try and figure out what you're
missing.

# How it works

<img src="https://raw.githubusercontent.com/catchouli/better_tactics/develop/screenshots/puzzle_complete.png">

When you complete a puzzle, the difficulty score you select will be used to calculate how long it should be until you see the puzzle again, and also to update your rating. The scores are interpreted as follows:

- 'Again' means that you failed to solve the puzzle on the first and need to review it again in the near future, in which case the Spaced Repetition algorithm will set the puzzle to be 're-learned', and you'll see it again the same day.
- 'Hard' meaning that you found the solution but that it was quite challenging, in which case you'll see the puzzle again in about the same amount of time since you last saw it.
- 'Good' is the neutral answer for a successful review, and the one that should be used primarily. After picking it the puzzle's review interval will be increased, and you'll see it less and less frequently over time.
- 'Easy' indicates you didn't find the puzzle very challenging at all, and is a good hint to the algorithm that you don't need to see it again very soon at all, and will cause its review interval to increase significantly.

The review button for each difficulty shows you the amount of time until you'll see that puzzle again if you pick that difficulty.

The difficulty you select is also used to calculate you a rating, according to the difficulty level of the puzzle, and how difficult you found it. 'Good' reviews will cause your rating to grow slowly over time, while 'Again' or 'Easy' reviews may cause larger swings in your rating. Initially, the algorithm will be very uncertain about your rating, and you may experience large swings, but this allows it to quickly find the right rating level for you as it becomes more and more accurate with each puzzle you complete. The rating algorithm used is <a href="https://en.wikipedia.org/wiki/Glicko_rating_system#Glicko-2_algorithm">Glicko2</a>, a common rating system for online chess and competitive games.

# Initial rating

The initial rating is currently 500. Because we use glicko2 ratings, it should increase pretty fast if you press 'easy' on puzzles beyond your current rating, but you may still find it takes a while for your rating to become accurate. A planned feature is some kind of rating wizard to set the initial rating more accurately on a per-user basis.

if you find you need to manually reset your rating or set it to a particular value, you can set it using the debug endpoint `/api/user/reset_rating/{desired_rating}`, which also resets your rating variance and should allow the app to re-find your rating level at about the given level. (e.g. <a href="http://localhost:3030/api/user/reset_rating/1500">http://localhost:3030/api/user/reset_rating/1500</a>)

# Acknowledgements

Made using <a href="https://www.rust-lang.org/">Rust</a>, <a href="https://github.com/seanmonstar/warp">warp</a>, and <a href="https://github.com/djc/askama">askama</a>. The Spaced Repetition algorithm used is the <a href="https://super-memory.com/english/ol/sm2.htm">SuperMemo 2 Algorithm</a>.

The puzzles are sourced from the <a href="https://database.lichess.org/#puzzles">lichess puzzles database</a>, which is amazing. Thanks, Thibault and the lichess community!

The chess board is also lichess's open source <a href="https://github.com/lichess-org/chessground">chessground</a> chess board
component, and the legal move detection and uses <a href="https://github.com/jhlywa/chess.js">chess.js</a>.

# Warning

This project is pretty functional but still experimental. You might encounter issues that require you to reset (or manually repair) your database. If you have trouble, let me know about it by making an issue, and hopefully I can fix it so it doesn't happen again (and if you send your database to me somehow I can try and repair it for you).


CHANGELOG.md
[1.11.0] - ????-??-??

Important: if you're building yourself, (i.e. not with the provided Dockerfile), npm is now needed
           in order to build the frontend. https://nodejs.org/en/download

Changed:
* The default rating variation on the 'Next puzzle' page is now 5% below the user's rating level,
  but not above the user's rating level anymore, as there was too much of a risk of showing puzzles
  that are too hard, especially as the user's rating level increased. There's now configuration
  values that can be changed to override this behavior if you preferred the previous +-5%.

[1.10.0] - 2023-11-02

Added:
* A hint button that shows a hint for the next move, for if the user can't work it out, but doesn't
  want to (or can't) go to lichess' analysis board to check the engine line. The puzzle will
  automatically be set to failed if the hint is used, but the user can always override it on the
  'Puzzle complete' panel and submit a positive response if desired.

Changed:
* Ctrl-c now causes a graceful shutdown and commits the database journal before closing it.
* The default interval for cards is now 1 minute. This doesn't affect much, but it makes it so that
  initially the 'Hard' and 'Good' buttons do something different (i.e. 1 minute and 10 minutes
  respectively), and so that pressing 'Again' on a puzzle makes it come up again in 1 minute
  instead of 10.

Fixed:
* A small fix to the scroll wheel behavior on the puzzle board to prevent the page from scrolling
  if the mousewheel is over the board.
* Fix countdown timer to next review not starting sometimes.
* The 'next review due' time on the stats screen no longer says 'never' when there are no reviews
  due for the rest of the day.

[1.9.0] - 2023-10-26

Added:
* Add simple move seeking (back and forward buttons) that allows you to see the moves played
  previously, and try playing out variations once the puzzle is over.
* Enable free movement after the wrong move is played the same as when the puzzle is complete,
  so you can play out variations and see what would have happened.
* Add config option to change the review order (SRS_REVIEW_ORDER) with three options:
  * DueTime (the current mode)
  * PuzzleRating (easier puzzles first)
  * Random (just show reviews in a random order)
* The default review order is now 'PuzzleRating', but the old behavior can be restored by using
  the option 'DueTime'.

Changed:
* Change initial computer move delay to 500, and make it and the subsequent move delays
  configurable with the configuration variables UI_INITIAL_MOVE_DELAY and UI_SUBSEQUENT_MOVE_DELAY.

[1.8.0] - 2023-10-21

Added:
* A Puzzle History page, which shows the history of all puzzles reviewed.
* On the 'next puzzle' screen, the puzzle to be reviewed is now saved until it is skipped.
* Skip buttons have been added to the puzzle interface for new puzzles:
  * "Skip this puzzle" when the puzzle is first started, which skips the current puzzle completely
    but doesn't award you any rating points.
  * "Too hard" and "Too easy" buttons, which also skip the puzzle, but correct your rating
    accordingly so you'll see easier or harder puzzles in the future.
  * "Don't repeat this puzzle", which marks the puzzle as complete, but without adding it to spaced
    repetition.
* Automated daily backups are now supported, and can be configured with the BACKUP_ENABLED,
  BACKUP_PATH, and BACKUP_HOUR configuration variables. See CONFIG.md for more details. The backup
  files are functional .sqlite databases that can be loaded back in as the application database,
  but do not include the puzzles, so if they are loaded the puzzles will be reinitialised when the
  application starts.

Removed:
* The configuration variable SQLITE_DB_NAME is now deprecated in favor of DATABASE_URL. If you
  have a SQLITE_DB_NAME set but not a DATABASE_URL, for compatibility DATABASE_URL will be set
  to "sqlite://{SQLITE_DB_NAME}". See CONFIG.md for more details.

Changed:
* Puzzles on the 'next puzzle' page will never be ones you've seen before anymore.
* Changes the .sqlite database to use WAL journaling (write-ahead-logging) so that multiple readers
  can use the database while something is writing to it. Overall the application should be a little
  faster.
* Tweaks the initial puzzle import to be a bit faster, and not lock the database as much. This is
  especially useful in docker, where the initial import was quite slow for me previously, and the
  web app was very slow to load while it was importing. The database import is now a lot faster,
  and page loads are generally instant while it is in progress. API calls that need to write (such
  as reviewing and skipping puzzles) might still need a couple of seconds while puzzles are being
  written to a database but it's much faster overall.

[1.7.0] - 2023-10-19

Changed:
* The puzzle page now loads the next puzzle automatically without refreshing the page.
* Cards leaving learning immediately due to reviewing them as 'Easy' now have their interval set
  to 4 days instead of the previous 1 day. This should allow actually easy puzzles to be seen far
  less frequently.
* The review forecast on the stats screen now allows you to change the number of days shown, and
  review scores allows you to see the data by percentage as well as total value.
* A link to the original lichess puzzle is now included on the puzzle interface.

[1.6.0] - 2023-10-13

Added:
* A review forecast on the stats page which shows the number of reviews due per day for the next
  week.
* A rating history which shows the user's rating growth over time.
* A review score histogram which shows the review scores being picked by the user by puzzle rating.

Changed:
* The user's rating deviation is now hidden from the stats page if it's under 100, the threshold
  under which the user's rating is no longer provisional and should be quite accurate.

[1.5.2] - 2023-10-09

Changed:

src/app.rs
pub mod backup;

use std::env::{self, VarError};
use std::error::Error;
use std::fmt::Debug;
use std::net::{IpAddr, Ipv4Addr};
use std::str::FromStr;

use chrono::{NaiveTime, DateTime, Local, Duration};
use url::Url;

use crate::db::PuzzleDatabase;
use crate::services::tactics_service::TacticsService;
use crate::services::user_service::UserService;
use crate::srs::{SrsConfig, ReviewOrder};

/// The application useragent, e.g. "better_tactics/0.0.1".
pub static APP_USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"));

/// A type for database urls that implements Debug so that we can pretty print the app config
/// without the URLs printing out all their fields.
#[derive(Clone)]
pub struct DatabaseUrl(pub Url);

impl Debug for DatabaseUrl {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let DatabaseUrl(url) = self;
        write!(f, "\"{url}\"")
    }
}

/// The application configuration.
#[derive(Debug, Clone)]
pub struct AppConfig {
    pub bind_interface: IpAddr,
    pub bind_port: u16,
    pub database_url: DatabaseUrl,
    pub tactics: TacticsConfig,
    pub srs: SrsConfig,
    pub backup: BackupConfig,
    pub ui: UiConfig,
}

#[derive(Debug, Clone)]
pub struct TacticsConfig {
    pub puzzle_rating_variation_up: f32,
    pub puzzle_rating_variation_down: f32,
}

#[derive(Debug, Clone)]
pub struct BackupConfig {
    pub enabled: bool,
    pub path: String,
    pub hour: NaiveTime,
}

#[derive(Debug, Clone)]
pub struct UiConfig {
    pub initial_move_delay: u32,
    pub subsequent_move_delay: u32,
}

impl Default for AppConfig {
    fn default() -> Self {
        Self {
            bind_interface: IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)),
            bind_port: 3030,
            database_url: DatabaseUrl(Url::parse("sqlite://puzzles.sqlite")
                .expect("Failed to parse default database_url")),
            tactics: TacticsConfig::default(),
            srs: SrsConfig::default(),
            backup: BackupConfig::default(),
            ui: UiConfig::default(),
        }
    }
}

impl Default for TacticsConfig {
    fn default() -> Self {
        Self {
            puzzle_rating_variation_up: 0.0,
            puzzle_rating_variation_down: 0.05,
        }
    }
}

impl Default for BackupConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            path: "./backups".into(),
            hour: NaiveTime::from_hms_opt(4, 0, 0)
                .expect("Failed to parse default backup hour"),
        }
    }
}

impl Default for UiConfig {
    fn default() -> Self {
        Self {
            initial_move_delay: 500,
            subsequent_move_delay: 250,
        }
    }
}

impl AppConfig {
    /// Load the app config from a .env file or environment variables.
    pub fn from_env() -> Result<AppConfig, Box<dyn Error>> {
        let _ = dotenvy::dotenv();

        let defaults: AppConfig = Default::default();

        Ok(Self {
            bind_interface: Self::env_var("BIND_INTERFACE")?.unwrap_or(defaults.bind_interface),
            bind_port: Self::env_var("BIND_PORT")?.unwrap_or(defaults.bind_port),
            database_url: Self::get_database_url()?.unwrap_or(defaults.database_url),
            srs: SrsConfig {
                default_ease: Self::env_var("SRS_DEFAULT_EASE")?.unwrap_or(defaults.srs.default_ease),
                minimum_ease: Self::env_var("SRS_MINIMUM_EASE")?.unwrap_or(defaults.srs.minimum_ease),
                easy_bonus: Self::env_var("SRS_EASY_BONUS")?.unwrap_or(defaults.srs.easy_bonus),
                day_end_hour: Self::env_var::<u32>("SRS_DAY_END_HOUR")?
                    .map(|day_end_hour| NaiveTime::from_hms_opt(day_end_hour, 0, 0)
                            .ok_or_else(|| format!("Invalid srs day_end_hour {}", day_end_hour)))
                    .transpose()?
                    .unwrap_or(defaults.srs.day_end_hour),
                review_order: Self::env_var::<ReviewOrder>("SRS_REVIEW_ORDER")
                    .map_err(|e| format!("{e}, possible values: {}", ReviewOrder::possible_values()))?
                    .unwrap_or(defaults.srs.review_order),
            },
            tactics: TacticsConfig {
                puzzle_rating_variation_up: Self::env_var("TACTICS_PUZZLE_RATING_VARIATION_UP")?
                    .unwrap_or(defaults.tactics.puzzle_rating_variation_up),
                puzzle_rating_variation_down: Self::env_var("TACTICS_PUZZLE_RATING_VARIATION_DOWN")?
                    .unwrap_or(defaults.tactics.puzzle_rating_variation_down),
            },
            backup: BackupConfig {
                enabled: Self::env_var("BACKUP_ENABLED")?.unwrap_or(defaults.backup.enabled),
                path: Self::env_var("BACKUP_PATH")?.unwrap_or(defaults.backup.path),
                hour: Self::env_var::<u32>("BACKUP_HOUR")?
                    .map(|day_end_hour| NaiveTime::from_hms_opt(day_end_hour, 0, 0)
                            .ok_or_else(|| format!("Invalid backup hour {}", day_end_hour)))
                    .transpose()?
                    .unwrap_or(defaults.backup.hour),
            },
            ui: UiConfig {
                initial_move_delay: Self::env_var("UI_INITIAL_MOVE_DELAY")?
                    .unwrap_or(defaults.ui.initial_move_delay),
                subsequent_move_delay: Self::env_var("UI_SUBSEQUENT_MOVE_DELAY")?
                    .unwrap_or(defaults.ui.subsequent_move_delay),
            },
        })
    }

    /// Get the database address from environment variables.
    pub fn get_database_url() -> Result<Option<DatabaseUrl>, Box<dyn Error>> {
        let db_url = Self::env_var("DATABASE_URL")?
            // Support old DB_NAME variable.
            .or(Self::env_var("SQLITE_DB_NAME")?
                .map(|db_name: String| format!("sqlite://{db_name}")))
            // Parse to URL.
            .map(|s| Ok(DatabaseUrl(Url::parse(&s)?)) as Result<_, Box<dyn Error>>)
            .transpose()?;

        Ok(db_url)
    }

    /// Read an env var and get a value, attempting to parse it to the specified type. If the
    /// variable is not set, returns the specified default.
    fn env_var<T>(key: &str) -> Result<Option<T>, Box<dyn Error>>
    where
        T: FromStr,
        <T as FromStr>::Err: Error + 'static,
    {
        match env::var(key) {
            Ok(value) => {
                let parsed = value.parse()
                    .map_err(|err| format!("Error parsing {key}: {err}"))?;
                Ok(Some(parsed))
            },
            Err(VarError::NotPresent) => Ok(None),
            Err(err) => Err(format!("Var error {key}: {err}").into())
        }
    }

    /// Get the last backup time.
    pub fn last_backup_datetime(&self) -> DateTime<Local> {
        crate::util::next_time_after(Local::now(), self.backup.hour) - Duration::days(1)
    }
}

/// The application state.
#[derive(Clone)]
pub struct AppState {
    pub app_config: AppConfig,
    pub user_service: UserService,
    pub tactics_service: TacticsService,
}

impl AppState {
    pub fn new(app_config: AppConfig, db: PuzzleDatabase) -> AppState {
        Self {
            user_service: UserService::new(app_config.clone(), db.clone()),
            tactics_service: TacticsService::new(app_config.clone(), db.clone()),
            app_config,
        }
    }
}
